File: backend/src/main/java/coursepick/coursepick/domain/course/Meter.java
package coursepick.coursepick.domain.course;

import coursepick.coursepick.application.exception.ErrorType;

public record Meter(
        double value
) {
    public static Meter zero() {
        return new Meter(0);
    }

    public static Meter max() {
        return new Meter(Double.MAX_VALUE);
    }

    public Meter add(Meter other) {
        return new Meter(this.value() + other.value());
    }

    public Meter clamp(double min, double max) {
        return new Meter(Math.clamp(this.value(), min, max));
    }

    public boolean isWithin(Meter other) {
        return this.value() <= other.value();
    }

    public double getRateOf(Meter other) {
        if (this.value == 0) {
            throw ErrorType.INVALID_RATIO_BASE.create();
        }
        return other.value() / this.value();
    }
}

File: android/app/src/main/java/io/coursepick/coursepick/domain/course/Meter.kt
package io.coursepick.coursepick.domain.course

@JvmInline
value class Meter(
    val value: Double,
) : Comparable<Meter> {
    constructor(value: Int) : this(value.toDouble())

    override fun compareTo(other: Meter): Int = this.value.compareTo(other.value)

    operator fun compareTo(other: Int): Int = this.value.compareTo(other)

    operator fun compareTo(other: Double): Int = this.value.compareTo(other)

    companion object {
        const val METRIC_MULTIPLIER = 1

        val MAX_VALUE = Meter(Double.MAX_VALUE)
    }
}
